src/main/java/net/neoforged/neoforge/event/entity/player/BlockInteractEvent.java

```diff
@@ -0,0 +1,108 @@
+import net.minecraft.world.item.Item;
+import net.minecraft.world.item.context.UseOnContext;
+import net.minecraft.world.level.block.state.BlockBehaviour;
+import net.neoforged.bus.api.Event;
+import net.neoforged.bus.api.ICancellableEvent;
+import net.neoforged.neoforge.common.extensions.IItemExtension;
+
+/**
+ * <p></p>Fires on both the client and server thread when a player uses an item targeting a block.
+ *
+ * <p>The event fires in three phases, corresponding with the three interaction behaviors:
+ * {@link IItemExtension#onItemUseFirst},
+ * {@link BlockBehaviour#use},
+ * and {@link Item#useOn}.</p>
+ *
+ * <p>If the event is cancelled via {@link #cancelWithResult},
+ * then the normal interaction behavior for that phase will not run,
+ * and the specified {@link InteractionResult} will be returned instead.</p>
+ */
+public class BlockInteractEvent extends PlayerInteractEvent implements ICancellableEvent
+{
+    private final UseOnContext context;
+    private final UsePhase usePhase;
+    private InteractionResult cancellationResult = InteractionResult.PASS;
+
+    public BlockInteractEvent(UseOnContext context, UsePhase usePhase)
+    {
+        super(context.getPlayer(), context.getHand(), context.getClickedPos(), context.getClickedFace());
+        this.context = context;
+        this.usePhase = usePhase;
+    }
+
+    /**
+     * @return context
+     */
+    public UseOnContext getUseOnContext()
+    {
+        return this.context;
+    }
+
+    /**
+     * @return The Use Phase of the interaction, see {@link UsePhase} for semantics
+     */
+    public UsePhase getUsePhase()
+    {
+        return this.usePhase;
+    }
+
+    /**
+     * <p>Cancels the use interaction (preventing the block or item's use behavior from running) and provides the
+     * specified result to the interaction logic instead.</p>
+     *
+     * <p>Invoke this if you intend to prevent the default interaction behavior and replace it with your own.</p>
+     *
+     * @param result The interaction result to return to the interaction logic.
+     * <ul>
+     * <li>SUCCESS, CONSUME, CONSUME_PARTIAL, and FAIL will prevent further types of interaction attempts
+     * when provided from the ITEM_BEFORE_BLOCK phase.
+     * <li>SUCCESS, CONSUME, and CONSUME_PARTIAL will trigger advancements on the server (except in the ITEM_BEFORE_BLOCK phase),
+     * and will also prevent the ITEM_AFTER_BLOCK item interaction from occurring if provided during the BLOCK phase.
+     * <li>SUCCESS will trigger the arm-swinging mechanic.
+     * <li>PASS will always allow proceeding to the next phase.
+     * </ul>
+     */
+    public void cancelWithResult(InteractionResult result)
+    {
+        this.cancellationResult = result;
+        this.setCanceled(true);
+    }
+
+    /**
+     * @return The result that the canceller cancelled this event with.<br>
+     * Not meaningful unless and until this event has been cancelled; intended to be
+     * used by the event firer.
```

Comment:
```suggestion
     * {@return the result that the canceller cancelled this event with}
     * This value is not meaningful unless and until this event has been cancelled; intended to be
     * used by the event firer.
```

Reply:
Mooted, remembered that PlayerInteractEvent has its own cancellation result stuff, removed this method